cloud/aws/templates/aws_oidc/bin/aws_cli.py

```diff
@@ -50,32 +51,70 @@ def restart_ecs_service(self):
             f"ecs update-service --force-new-deployment --cluster={self._ecs_cluster} --service={self._ecs_service}"
         )
 
-    def wait_for_ecs_service_healthy(self):
+    def wait_for_ecs_service_healthy(self, previous_deployment_id):
         """
         Polls the CiviForm ECS service, waiting for the PRIMARY deployment to
-        have rolloutStatus of COMPLETED.
+        have rolloutStatus of COMPLETED. If the PRIMARY deployment ID ends up
+        being equal to the previous_deployment_id, then the deployment failed
+        and we've rolled back.
 
         Gives up after 60 tries, sleeps 30 seconds between each try.
         """
         print(
             "\nWaiting for CiviForm ECS service to become healthy.\n"
             f"Service URL: {self._get_url_of_ecs_service()}")
 
+        error_text = inspect.cleandoc(
+            """
+            go to the Service URL printed above and click on the logs tab.
+            More details at https://docs.civiform.us/it-manual/sre-playbook/terraform-deploy-system/terraform-aws-deployment#inspecting-logs
+            For debugging help, contact the CiviForm oncall: https://docs.civiform.us/governance-and-management/project-management/on-call-guide#on-call-responsibilities
+            """)
+
+        tracked_deployment_id = None
         tries = 60
         while True:
-            state = self._ecs_service_state()
+            info = self.ecs_service_state()
+            id = info["id"]
+            state = info["state"]
+            tracked_deployment_id = id if tracked_deployment_id is None else tracked_deployment_id
+            if id == previous_deployment_id:
+                print(
+                    "ERROR: It appears that the deployment has failed and we've rolled back to the previous working deployment.\n"
+                    "This usually means the new tasks are crash-looping and we've rolled back to the previous deployment.\n"
+                    "To view the logs to see what happened, " + error_text)
+                raise Exception("deployment rolled back")
+
             if state == "COMPLETED":
+                # If the ID of the deployment that is COMPLETED is not the one we've been
+                # waiting for, then:
+                #   - This is a deployment that bypassed the "deployment is already in progress"
```

Comment:
question (non-blocking): This list is also confusing to me, but I can't tell if it's because I'm confused by the wording or by the possible scenarios. If this stays in, let's talk through it together.